src/registry/AvatarRegistry.sol

```diff
@@ -0,0 +1,541 @@
+            revert NotMinLinkFundedUpKeep();
+        }
+        if (LINK.balanceOf(address(this)) < linkAmount) {
+            _swapEthForLink(linkAmount);
+        }
+
+        /// @dev check registry state before registering
+        (IKeeperRegistry.State memory state,,) = CL_REGISTRY.getState();
+        uint256 oldNonce = state.nonce;
+
+        bytes memory data = abi.encodeCall(
+            IKeeperRegistrar.register,
+            (
+                name,
+                bytes(""),
+                targetAddress,
+                uint32(gasLimit),
+                address(this),
+                bytes(""),
+                uint96(linkAmount),
+                0,
+                address(this)
+            )
+        );
+
+        LINK.transferAndCall(KEEPER_REGISTRAR, linkAmount, data);
+
+        (state,,) = CL_REGISTRY.getState();
+        uint256 newNonce = state.nonce;
+
+        if (newNonce == oldNonce + 1) {
+            upkeepID = uint256(
+                keccak256(abi.encodePacked(blockhash(block.number - 1), address(CL_REGISTRY), uint32(oldNonce)))
+            );
+        } else {
+            revert NotAutoApproveKeeper();
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // PUBLIC VIEW
+    ////////////////////////////////////////////////////////////////////////////
+
+    /// @dev Runs off-chain at every block to determine if the `performUpkeep`
+    /// function should be called on-chain.
+    function checkUpkeep(bytes calldata)
+        external
+        view
+        override
+        whenNotPaused
+        returns (bool upkeepNeeded_, bytes memory performData_)
+    {
+        address[] memory avatarsInTestStatus = getAvatarsByStatus(AvatarStatus.TESTING);
+        bool underFunded;
+
+        uint256 avatarsTestLength = avatarsInTestStatus.length;
+        if (avatarsTestLength > 0) {
+            /// @dev loop thru avatar in test status for register or topup if required
+            for (uint256 i = 0; i < avatarsTestLength; i++) {
+                /// @dev requires that CL keeper is config properly
+                if (IAvatar(avatarsInTestStatus[i]).keeper() != address(CL_REGISTRY)) {
+                    continue;
+                }
+
+                /// @dev prio `test` avatar unregister
+                if (avatarsInfo[avatarsInTestStatus[i]].upKeepId == 0) {
+                    upkeepNeeded_ = true;
+                    performData_ = abi.encode(avatarsInTestStatus[i], OperationKeeperType.REGISTER_UPKEEP);
+                    break;
+                }
+
+                /// @dev check for under funded avatar upkeeps
+                (,, underFunded) = _isAvatarUpKeepUnderFunded(avatarsInTestStatus[i]);
+                if (underFunded) {
+                    upkeepNeeded_ = true;
+                    performData_ = abi.encode(avatarsInTestStatus[i], OperationKeeperType.TOPUP_UPKEEP);
+                    break;
+                }
```

Comment:
Does this ignore cancelled upkeeps?

Reply:
it does, since it will check all the avatars now, regardless of their status via `getAvatars`, which only those active appears, since in the cancelled ones are removed from the set

Reply:
But `cancelAvatarUpKeep` doesn't remove it from the set
https://github.com/Badger-Finance/badger-avatars/blob/ae6121947c70e8639e9edb448803be45dec4517f/src/registry/AvatarRegistry.sol#L194-L204

Reply:
added tighter checks on checkupkeep and upkeep for those members which may be under cancellation process in 580e830b80c3866c0671300d931f95ad9c3e8b10 to jump over them in checkupkeep and in performupkeep revert in case that the specific upkeep id is found to be in cancellation mode